design-proposals/klip-1-improve-udf-interfaces.md

```diff
@@ -0,0 +1,314 @@
+struct support, generic types, variable arguments and complex aggregation*
+
+
+## Motivation and background
+
+There have been many requests ([#2163](https://github.com/confluentinc/ksql/issues/2163)) to
+improve the UDF/UDAF interface in KSQL. Of these features, four stand out:
+
+- _Custom Struct Support_ - UDFs can only operate on standard Java types (`Map`, `List`, `String`, 
+etc...) and do not accept structured types. This dramatically limits what KSQL can handle, as
+much of the data flowing through kafka is structured Avro/JSON.
+- _Generics Type Support_ - Today, there is no way to implement a single UDF that supports data 
+structures with implicit types. For example, implementing a generic `list_union` operator would
+require different method signatures for each supported type. This quickly becomes unmanageable
+when there are multiple arguments.
+- _Variable Argument Support_ - Varargs are required to create UDFs that accept an arbitrary
+number of inputs. Creating a simple `concat` UDF that could sum multiple columns would require
+creating method signatures for as many input combinations as desired.
+- _Complex Aggregation Models_ - Today, KSQL only supports tracking a single variable for 
+aggregation and it must mirror the output type. To implement an `avg` UDAF, it is necessary to store 
+both a `count` and a `running_total`.
+
+## Scope
+
+All changes will be made for the new annotation-based UDF interfaces and will not be backported to
+the legacy implementation. The bullet points mentioned above are all in scope, but the following 
+will not be included in this KLIP:
+
+- UDTFs - multiple output operations are more complicated and will require a KLIP of their own
+- Generics/Varargs for UDAFs - the support for Generics/Varargs will only extend to UDFs.
+
+## Value
+
+Basic UDF operations that are supported in SQL languages will now be unblocked for KSQL. Some of
+these are outlined below:
+
+|      Improvement      |                                 Value                                  |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | Unblock `ARRAY_LENGTH`, `ARRAY_SLICE`, `ARRAY_TO_STRING`, etc...       |
+| Varargs support       | `CONCAT`, `MAX`, `MIN` etc...                                          |
+| Complex Aggregation   | `AVG`, `PERCENTILE`, etc...                                            |
+| Structured UDFs       | Arbitrary data handling without intermediary transforms                |
+
+## Public APIS
+
+Some public APIs will be changed, though all will be changed in a backwards compatible fashion:
+
+|      Improvement      |                               API change                               |
+|:---------------------:|------------------------------------------------------------------------|
+| Generics support      | UDF interface will accept inferred generics (e.g. `List<T>`)           |
```

Comment:
**Signature for generic functions**: It looks like neither Hive nor most other SQL languages support this type of functionality (Hive has it through the `ObjectInspector` interface, which is quite bulky...). I think we have two options here:

* Generate all possible signatures from a list of all supported types (e.g. `DESCRIBE FUNCTION` would should `ARRAY<VARCHAR>`, `ARRAY<INT>`, `ARRAY<BIGINT>`...). This would allow us to punt the problem to another time and instead "simulate" registering all possible generics. This can get super complicated when you infer more than one type (obviously it's exponential complexity), but we can likely limit it to one generic inference for now and cover 80% of use cases.
* Introduce a new syntax for generics. Off the top of my head, I imagine that it would looks something like `ARRAY<?>`.

**Support for non-container generics**: Yes. We should definitely support this, and the same strategy should apply (e.g. we must be able to infer the output schema from the input parameters) and we would resolve method signatures in the same was described above.

Reply:
Talked with the team and we believe that the first alternative isn't a good option. The most updated revision has the up-to-date recommendation, which will be to increment a type `T1`...`TN` for each generic inferred type.